src/sysext/sysext.c

```diff
@@ -525,35 +692,79 @@ static int merge_hierarchy(
                 const char *meta_path,
                 const char *overlay_path) {
 
-        _cleanup_free_ char *resolved_hierarchy = NULL, *f = NULL, *buf = NULL;
+        _cleanup_free_ char *resolved_hierarchy = NULL, *f = NULL, *buf = NULL, *resolved_upper_dir = NULL;
         _cleanup_strv_free_ char **layers = NULL;
         struct stat st;
         int r;
+        bool drop_hierarchy = false;
 
         assert(hierarchy);
         assert(meta_path);
         assert(overlay_path);
 
+        f = extension_data_path_for_hierarchy(hierarchy);
+        if (!f)
+                return log_oom();
+
+        r = chase(f, arg_root, CHASE_PREFIX_ROOT, &resolved_upper_dir, NULL);
+        if (r == -ENOENT)
+                log_debug_errno(r, "Hierarchy '%s' on host has no directory %s for storing writes, merging as read-only.", hierarchy, f);
+        else if (r < 0)
+                return log_error_errno(r, "Failed to resolve directory for storing writes on host hierarchy '%s': %m", hierarchy);
+
         /* Resolve the path of the host's version of the hierarchy, i.e. what we want to use as lowest layer
          * in the overlayfs stack. */
         r = chase(hierarchy, arg_root, CHASE_PREFIX_ROOT, &resolved_hierarchy, NULL);
         if (r == -ENOENT)
                 log_debug_errno(r, "Hierarchy '%s' on host doesn't exist, not merging.", hierarchy);
         else if (r < 0)
                 return log_error_errno(r, "Failed to resolve host hierarchy '%s': %m", hierarchy);
-        else {
+        /* Actually an if, but using while to jump out of the block easily. */
```

Comment:
uh. mayb split this out in a helper instead?

Reply:
Done.